### PR TITLE
rework choice providers

### PR DESCRIPTION
**Pull request by dannyroberts**
to provide more related functionality as a class with a shared ABC
rather than having them be a single function

@czue I'm cherry picking this from https://github.com/dimagi/commcare-hq/pull/9336 to run tests, because it's a reasonable change on its own, and because it's a little more invasive/refactory than the rest of that PR. (The other PR depends on this though.)

### CODE REVIEW

File: corehq/apps/userreports/reports/filters/factory.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def _build_dynamic_choice_list_filter(spec):
         label=wrapped.display,
         show_all=wrapped.show_all,
         url_generator=dynamic_choice_list_url,
-        choice_provider=FilterChoiceProviderFactory.from_spec(wrapped.get_choice_provider_spec()),
+        choice_provider=FilterChoiceProviderFactory.from_spec(wrapped.get_choice_provider_spec())(report, wrapped.slug),
```
Comment (dimagimon) [inline corehq/apps/userreports/reports/filters/factory.py]: E501 line too long (120 > 115 characters)

**Comment by czue** [comment]
looks good!

File: corehq/apps/userreports/reports/filters/choice_providers.py [hunk 2]
```diff
@@ -25,34 +23,77 @@ def offset(self):
...
+
+class LocationChoiceProvider(ChoiceProvider):
+
+    def query(self, query_context):
+        # todo: consider making this an extensions framework similar to custom expressions
+        # todo: does this need fancier permission restrictions and what not?
+        # see e.g. locations.views.child_locations_for_select2
+        if query_context.query:
+            locations = SQLLocation.active_objects.filter_path_by_user_input(
+                domain=query_context.domain, user_input=query_context.query
```
Comment (dannyroberts) [inline corehq/apps/userreports/reports/filters/choice_providers.py]: I noticed while coding other stuff that I made an error refactoring which would have broken this (but I looking through email I'm not immediately finding any 500s yet). Here's the hotfix: https://github.com/dimagi/commcare-hq/pull/9337

Comment (dannyroberts) [inline corehq/apps/userreports/reports/filters/choice_providers.py]: I noticed while coding other stuff that I made an error refactoring which would have broken this (but I looking through email I'm not immediately finding any 500s yet). Here's the hotfix: https://github.com/dimagi/commcare-hq/pull/9357

Comment (czue) [inline corehq/apps/userreports/reports/filters/choice_providers.py]: thanks. I think the icds supervisor app is the only thing that uses this so should be ok.

**Pull request merged**